libcloud/compute/drivers/scaleway.py

```diff
@@ -0,0 +1,295 @@
+
+
+class ScalewayConnection(ConnectionUserAndKey):
+    """
+    Connection class for the Scaleway driver.
+    """
+
+    host = SCALEWAY_API_HOST
+    allow_insecure = False
+    responseCls = ScalewayResponse
+
+    def add_default_headers(self, headers):
+        """
+        Add headers that are necessary for every request
+        """
+        headers['X-Auth-Token'] = self.key
+        headers['Content-Type'] = 'application/json'
+        return headers
+
+
+def _kb_to_mb(size):
+    return int(size / 1000 / 1000 / 1000)
+
+
+def _mb_to_kb(size):
+    return int(size * 1000 * 1000 * 1000)
+
+
+class ScalewayNodeDriver(NodeDriver):
+    """
+    Scaleway NodeDriver
+    """
+
+    type = Provider.SCALEWAY
+    connectionCls = ScalewayConnection
+    name = 'Scaleway'
+    website = 'https://www.scaleway.com/'
+
+    SNAPSHOT_STATE_MAP = {  # TODO map all states
+        'snapshotting': VolumeSnapshotState.CREATING
+    }
+
+    def list_sizes(self):
+        return [NodeSize(driver=self, **copy.deepcopy(size))
+                for size in SCALEWAY_INSTANCE_TYPES]
+
+    def list_images(self):
+        response = self.connection.request('/images')
+        images = response.object['images']
+        return [NodeImage(id=image['id'],
+                          name=image['name'],
+                          driver=self,
+                          extra={'arch': image['arch']})
+                for image in images]
+
+    def create_image(self, node, name, description=None):
+        data = {
+            'organization': self.key,
+            'name': name,
+            'arch': node.extra['arch'],
+            'root_volume': node.extra['volumes']['0']['id']  # TODO check this
+        }
+        response = self.connection.request('/images', data=json.dumps(data),
+                                           method='POST')
+        image = response.object['image']
+        return NodeImage(id=image['id'], name=image['name'], driver=self)
+
+    def delete_image(self, node_image):
+        return self.connection.request('/images/%s' % node_image.id,
+                                       method='DELETE').success()
+
+    def get_image(self, image_id):
+        response = self.connection.request('/images/%s' % image_id)
+        image = response.object['image']
+        return NodeImage(id=image['id'], name=image['name'], driver=self)
+
+    def list_nodes(self):
+        response = self.connection.request('/servers')
+        servers = response.object['servers']
+        return [self._to_node(server) for server in servers]
+
+    def _to_node(self, server):
+        public_ip = server['public_ip']
+        private_ip = server['private_ip']
+        return Node(id=server['id'],
+                    name=server['name'],
+                    state=NodeState.fromstring(server['state']),
+                    public_ips=[public_ip['address']] if public_ip else [],
+                    private_ips=[private_ip['address']] if private_ip else [],
```

Comment:
This didn't (any more?) work. The returned structure looked like:
 ```json
      "public_ip": {
        "dynamic": false,
        "id": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
        "address": "A.B.C.D"
      },
      "private_ip": "N.O.P.Q",
```
So changing this line to `private_ips=[private_ip] if private_ip else [],` made everything work ok for me.